FF113 supports WebGPU on nightly

hamishwillee opened the pull request:
FF113 supports WebGPU on nightly in https://bugzilla.mozilla.org/show_bug.cgi?id=1746245

Spec has iterated a lots since this was defined. Probably not worth tracking introduction of various parts behind pref. This just tracks broad WebGPU support as Preview.

Open questions on devs here: https://bugzilla.mozilla.org/show_bug.cgi?id=1746245#c21

Other docs work can be tracked in https://github.com/mdn/content/issues/26157

--- 

List is what is documented and in BCD. X indicates also in WebIdl for FF. So this is check of coverage.

- [x] GPU
- [x] GPUAdapter
- [x] GPUAdapterInfo
- [x] GPUBindGroup
- [x] GPUBindGroupLayout
- [x] GPUBuffer
- [x] GPUCanvasContext
- [x] GPUCommandBuffer
- [x] GPUCommandEncoder
- [x] GPUCompilationInfo
- [x] GPUCompilationMessage
- [x] GPUComputePassEncoder - behind pref `dispatchWorkgroupsIndirect`
- [x] GPUComputePipeline - not yet serializable
- [x] GPUDevice - note, not support createQuerySet()
- [x] GPUDeviceLostInfo
- [ ] GPUError - not in IDL or  visible on console.
- [ ] GPUExternalTexture - not in IDL or  visible on console.
- [ ] GPUInternalError - not in IDL or  visible on console.
- [x] GPUOutOfMemoryError (constructor commented out in IDL but seems to inherit one)
- [ ] GPUPipelineError - not in IDL or  visible on console.
- [x] GPUPipelineLayout
- [x] GPUQuerySet 
- [x] GPUQueue - IDL comment out `onSubmittedWorkDone()`
- [x] GPURenderBundle
- [x] GPURenderBundleEncoder - Behind pref is `drawIndirect()`, `drawIndexedIndirect()`.
- [x] GPURenderPassEncoder - IDL comments out `beginOcclusionQuery`, `endOcclusionQuery`, `beginPipelineStatisticsQuery`, `endPipelineStatisticsQuery()`, `writeTimestamp()`. Also behind pref is `drawIndirect()`, `drawIndexedIndirect()`.
- [x] GPURenderPipeline - not yet serializable
- [x] GPUSampler
- [x] GPUShaderModule
- [x] GPUSupportedFeatures
- [x] GPUSupportedLimits
- [x] GPUTexture
- [x] GPUTextureView
- [x] GPUUncapturedErrorEvent
- [x] GPUValidationError - oddly the `message` is shown as in IDL for FF (not inherited) but as inherited in spec.
- [x] HTMLCanvasElement.getContext()
- [x] Navigator.gpu
- [ ] WorkerNavigator.gpu - NOT supported - so you can't get a handle on this in workers on FF

A few issues:
- GPUComputePipeline  and GPUrenderPipeline are serializable but that isn't supported yet. I don't think we need to track that in preview right? i.e. I assume when this goes to release it will be serializable and that is the only thing that matters. Of course hard to remember that needs to be confirmed.

Elchi3 commented on api/GPU.json:
```diff
@@ -14,7 +14,9 @@
+            "partial_implementation": true,
+            "notes": "Currently supported on Linux and Windows only. See <a href='https://bugzil.la/1616739'>bug 1616739</a>."
           },
           "firefox_android": "mirror",
```
Android needs version_false and shouldn't mirror.

hamishwillee replied:
Done everywhere except where there is an addition flag which means this is disabled by default anyway.

hamishwillee commented on api/GPUCanvasContext.json:
```diff
@@ -14,9 +14,13 @@
+            "partial_implementation": true,
+            "notes": "Currently supported on Linux and Windows only. See <a href='https://bugzil.la/1616739'>bug 1616739</a>."
+          },
+          "firefox_android": {
```
@Elchi3  item `canvas` below is in spec, but I can't see it in IDL. Am I missing something obvious in https://searchfox.org/mozilla-central/source/dom/webidl/WebGPU.webidl#1087

hamishwillee replied:
Ignore that. I think my brain broke.

hamishwillee commented on api/GPUQuerySet.json:
```diff
@@ -50,9 +54,13 @@
             },
             "edge": "mirror",
             "firefox": {
+              "version_added": "preview",
```
This `count` and `type` are in a dictionary `GPUQuerySetDescriptor`, and it is not so obvious these become properties of GPUQuerySet. Question is, if these are properties, why is `pipelineStatistics` not present - should it be?

chrisdavidmills replied:
Where are you getting `pipelineStatistics` from? I can't see it in the [editor's draft](https://gpuweb.github.io/gpuweb/) anywhere.

hamishwillee commented on api/GPUOutOfMemoryError.json:
```diff
@@ -14,9 +14,13 @@
             "version_added": false
           },
-          "firefox_android": "mirror",
           "ie": {
```
Note, the constructor subfeature is commented out in idl - which should mean "not implemented" - right?

chrisdavidmills replied:
Sounds like it to me. You could try testing it using the code snippet available at https://developer.mozilla.org/en-US/docs/Web/API/GPUOutOfMemoryError#examples

hamishwillee replied:
Will do, tomorrow. Thanks for the advice.

hamishwillee commented:
Thanks @Elchi3 - I've updated all the `firefox-android` records to false. I've also updated everything else I can see is in the Firefox IDL. I'm still waiting on confirmation of some things - IDL sometimes seems a bit magic to me, and the FF IDL doesn't seem to always match spec IDL.

A few things though
1. This is enabled by default in nightly in FF113, which is what I have recorded. 
   - I know it is behind a pref since before FF70 but not clear the precise version, or what versions various features were added. I might get that info from the dev team, but I might not. Now this is in preview, is it worth chasing this information about the pref? 
   - I'd rather not - from what I can tell if you want to play with this you can/should take nightly.
2. Some of the items are serializable in spec but not in FF. I am assuming that this will be fixed for release, so am not adding serializable subfeature to track those.

chrisdavidmills commented:
>  know it is behind a pref since before FF70 but not clear the precise version, or what versions various features were added. I might get that info from the dev team, but I might not. Now this is in preview, is it worth chasing this information about the pref?

@hamishwillee I don't think this would be very helpful anyway. The spec has changed quite a bit since then.

hamishwillee commented:
> > Some of the items are serializable in spec but not in FF. I am assuming that this will be fixed for release, so am not adding serializable subfeature to track those.
> 
> Should I be doing that for Chrome?

If chrome didn't implement all the interfaces then realistically I think you'd want to add a serializable subfeature to track what is and is not supported. But I am not sure whether it should be added to the ones that do support the feature or do not, or both! 

We need advice from @Elchi3 or @queengooborg

hamishwillee commented:
Hi @Elchi3 , @chrisdavidmills 

I've done some more research/testing today and update this appropriately. It is now as far as I can tell accurate and ready for review.

Note however that it may not yet be complete w.r.t. behaviour features. 
Specifically, I THINK from the data guidelines for APIs that we should record the support status for behaviour features, whether or not they are supported or have different compatibility in any browser: https://github.com/mdn/browser-compat-data/blob/main/docs/data-guidelines/api.md#secure-context-required-secure_context_required
If I am correct we need to add secure_context_required and worker_support  to any marked in the IDL as such. 
Further, serialization is also a (newer) behaviour but there is no data guideline. We should add one, and it should match the logic for the others. @Elchi3 can you confirm the rule?

Even so I would prefer we merged this and then added the behaviours as a follow on - OK?

Elchi3 commented:
> Further, serialization is also a (newer) behaviour but there is no data guideline. We should add one, and it should match the logic for the others. @Elchi3 can you confirm the rule?

I'm not quite sure I have the context here. Happy to review data and work on a guideline as a followup. Do you have pointers for me?

Elchi3 commented on api/GPU.json:
```diff
@@ -14,9 +14,13 @@
           "firefox": {
+            "version_added": "preview",
+            "partial_implementation": true,
+            "notes": "Currently supported on Linux and Windows only. See <a href='https://bugzil.la/1616739'>bug 1616739</a>."
```
If I care about macOS support, I can't really see anything in the linked bug. Is there a bug that tracks that specifically? Otherwise I think we should remove the link.

hamishwillee replied:
Can't find one. Stripping out the link.

hamishwillee commented:
@Elchi3 2. W.r.t. secure_context, fair enough. 

> I'm not quite sure I have the context here. Happy to review data and work on a guideline as a followup. Do you have pointers for me?

So a [serializable object](https://developer.mozilla.org/en-US/docs/Glossary/Serializable_object) is one that you can send in a postMessage or use in a structured clone. The list of things to which it applies is provided in [Structured clone algorithm > Supported types](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#supported_types). When I added Errors to that list, I also made sure their descriptions mentioned this feature (but did not add to others):

![image](https://user-images.githubusercontent.com/5368500/233509893-a87b3386-f132-4179-860c-78ca97a304f5.png)

The data rule is the same as for secure content - you add it if there is a compatibility change. That was the case for the errors - see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/EvalError#browser_compatibility

Is that enough?

MDN docs side I think we probably should add metadata for serializable, secure context, etc by default and use icons or whatever to indicate these "features".

hamishwillee commented:
@Elchi3 I removed the bugzilla link (agree not useful). Serialization has not been implemented on anything yet as far as we can tell, so there is nothing to record there. Secure context has been implemented as far as we can tell so compatibility issue to record.

Upshot, I think this is good and ready to go in. Can you review/merge?

Elchi3 commented:
Thanks Hamish! This all sounds great to me. Thanks for the info on serialization. Would be nice to formalize this in a guideline (on both BCD and MDN). How it is done for `EvalError` looks good to me.

Elchi3 reviewed (approved):
Looks good to me.
